### PR TITLE
Add keyword substitution for HTML components

### PR DESCRIPTION
**Pull request by kluo**
@stvstnfrd @caesar2164

### CODE REVIEW

File: common/lib/xmodule/xmodule/tests/test_html_module.py [hunk 1]
```diff
@@ -1,22 +1,33 @@
...
-class HtmlModuleSubstitutionTestCase(unittest.TestCase):
+class HtmlModuleSubstitutionTestCase(ModuleStoreTestCase):
     descriptor = Mock()
 
     def test_substitution_works(self):
         sample_xml = '''%%USER_ID%%'''
         field_data = DictFieldData({'data': sample_xml})
-        module_system = get_test_system()
+
+        user = UserFactory.create()
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_html_module.py]: looks like you can't get UserFactory in ```common/lib/xmodule/xmodule``` so you're gonna have to Mock the user? ```user=Mock(username='ma', email='ma@edx.org', is_staff=False, is_active=True),```

File: common/lib/xmodule/xmodule/js/src/html/edit.coffee [hunk 2]
```diff
@@ -990,6 +990,27 @@ class @HTMLEditingDescriptor
...
     return @visualEditor
 
+  _validateString = (string) ->
+    regexp = /%%[^%\s]+%%/g
+    keywordsSupported = [
+      '%%USER_ID%%',
+      '%%USER_FULLNAME%%',
+      '%%COURSE_DISPLAY_NAME%%',
+      '%%COURSE_START_DATE%%',
+      '%%COURSE_END_DATE%%'
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/js/src/html/edit.coffee]: I _think_ Coffeescript supports trailing commas.

File: common/lib/xmodule/xmodule/js/src/html/edit.coffee [hunk 3]
```diff
@@ -1001,4 +1022,11 @@ class @HTMLEditingDescriptor
     if text == undefined
       text = @advanced_editor.getValue()
 
+    validation = _validateString(text)
+    if not validation.isValid
+      message = gettext("There are invalid keywords in your email. Please check the following keywords and try again:")
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/js/src/html/edit.coffee]: Can we use single quotes for consistency?

File: common/lib/xmodule/xmodule/js/src/html/edit.coffee [hunk 4]
```diff
@@ -1001,4 +1022,11 @@ class @HTMLEditingDescriptor
...
+      message += "\n" + validation.keywordsInvalid.join('\n')
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/js/src/html/edit.coffee]: Can we use single quotes for consistency?

File: common/lib/xmodule/xmodule/js/src/html/edit.coffee [hunk 5]
```diff
@@ -1001,4 +1022,11 @@ class @HTMLEditingDescriptor
...
+      alert message
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/js/src/html/edit.coffee]: Can we use parens in the function call for consistency?

File: common/lib/xmodule/xmodule/tests/test_html_module.py [hunk 6]
```diff
@@ -14,25 +14,9 @@ class HtmlModuleSubstitutionTestCase(unittest.TestCase):
     def test_substitution_works(self):
         sample_xml = '''%%USER_ID%%'''
         field_data = DictFieldData({'data': sample_xml})
-        module_system = get_test_system()
-        module = HtmlModule(self.descriptor, module_system, field_data, Mock())
-        self.assertEqual(module.get_html(), str(module_system.anonymous_student_id))
-
-    def test_substitution_without_magic_string(self):
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_html_module.py]: Why don't we need this?

File: common/lib/xmodule/xmodule/tests/test_html_module.py [hunk 7]
```diff
@@ -14,25 +14,9 @@ class HtmlModuleSubstitutionTestCase(unittest.TestCase):
...
-                <p>Hi USER_ID!11!</p>
-            </html>
-        '''
-        field_data = DictFieldData({'data': sample_xml})
-        module_system = get_test_system()
-        module = HtmlModule(self.descriptor, module_system, field_data, Mock())
-        self.assertEqual(module.get_html(), sample_xml)
+        anon_id = '123456789'
 
-    def test_substitution_without_anonymous_student_id(self):
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_html_module.py]: Why don't we need this?

File: common/lib/xmodule/xmodule/tests/test_html_module.py (hunk 6 above)
Comment (kluo) [inline common/lib/xmodule/xmodule/tests/test_html_module.py]: The tests were based on the old substitution code, which only supported USER_ID

**Comment by stvstnfrd** [comment]
+1

**Pull request merged**